### PR TITLE
fix socket option setting for windows

### PR DESCRIPTION
**Pull request by qq53**
for windows, the SO_RCVTIMEO SO_RNDTIMEO parameter is DWORD, and in milliseconds

### CODE REVIEW

**Comment by yhirose** [comment]
@qq53, thanks for the pull request, but the code cannot be compiled. Could you take a look at it, and make sure that all the unit tests are passed with your change?

**Comment by qq53** [comment]
> @qq53, thanks for the pull request, but the code cannot be compiled. Could you take a look at it, and make sure that all the unit tests are passed with your change?

its my fault, i type wrong parameter name

**Comment by yhirose** [comment]
@qq53, thanks for your contribution!

File: httplib.h [hunk 1]
```diff
@@ -2687,16 +2687,27 @@ inline socket_t create_client_socket(
...
+#endif
         }
         {
-          timeval tv;
-          tv.tv_sec = static_cast<long>(write_timeout_sec);
-          tv.tv_usec = static_cast<decltype(tv.tv_usec)>(write_timeout_usec);
-          setsockopt(sock2, SOL_SOCKET, SO_SNDTIMEO, (char *)&tv, sizeof(tv));
+
+#ifdef _WIN32
+            uint32_t timeout = static_cast<uint32_t>(write_timeout_sec * 1000 + write_timeout_usec / 1000);
```
Comment (yhirose) [inline httplib.h:2703]: Should be `read_timeout_sec`

File: httplib.h [hunk 2]
```diff
@@ -5243,18 +5254,29 @@ inline bool Server::listen_internal() {
...
+            timeval tv;
+            tv.tv_sec = static_cast<long>(read_timeout_sec_);
+            tv.tv_usec = static_cast<decltype(tv.tv_usec)>(read_timeout_usec_);
+            setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, (char *)&tv, sizeof(tv));
+#endif
+        }
+        {
+
+#ifdef _WIN32
+            uint32_t timeout = static_cast<uint32_t>(write_timeout_sec_ * 1000 + write_timeout_usec_ / 1000);
```
Comment (yhirose) [inline httplib.h:5271]: Should be `read_timeout_sec`.

**Comment by yhirose** [comment]
I actually found more errors after I have merged the pull request. I just wonder if you really tested this code...

File: httplib.h [hunk 3]
```diff
@@ -5243,18 +5254,29 @@ inline bool Server::listen_internal() {
...
+            setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, (char *)&timeout, sizeof(timeout));
+#else
+            timeval tv;
+            tv.tv_sec = static_cast<long>(read_timeout_sec_);
+            tv.tv_usec = static_cast<decltype(tv.tv_usec)>(read_timeout_usec_);
```
Comment (yhirose) [inline httplib.h:5276]: Should use `write_timeout_sec`.

File: httplib.h [hunk 4]
```diff
@@ -2687,16 +2687,27 @@ inline socket_t create_client_socket(
...
-          tv.tv_sec = static_cast<long>(write_timeout_sec);
-          tv.tv_usec = static_cast<decltype(tv.tv_usec)>(write_timeout_usec);
-          setsockopt(sock2, SOL_SOCKET, SO_SNDTIMEO, (char *)&tv, sizeof(tv));
+
+#ifdef _WIN32
+            uint32_t timeout = static_cast<uint32_t>(write_timeout_sec * 1000 + write_timeout_usec / 1000);
+            setsockopt(sock2, SOL_SOCKET, SO_SNDTIMEO, (char *)&timeout, sizeof(timeout));
+#else
+            timeval tv;
+            tv.tv_sec = static_cast<long>(read_timeout_sec);
```
Comment (yhirose) [inline httplib.h:2707]: Should use `write_timeout_sec`.

File: httplib.h [hunk 5]
```diff
@@ -2687,16 +2687,27 @@ inline socket_t create_client_socket(
...
+            tv.tv_usec = static_cast<decltype(tv.tv_usec)>(read_timeout_usec);
```
Comment (yhirose) [inline httplib.h:2708]: Should use `write_timeout_usec`.

File: httplib.h [hunk 6]
```diff
@@ -5243,18 +5254,29 @@ inline bool Server::listen_internal() {
...
+#endif
+        }
+        {
+
+#ifdef _WIN32
+            uint32_t timeout = static_cast<uint32_t>(write_timeout_sec_ * 1000 + write_timeout_usec_ / 1000);
+            setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, (char *)&timeout, sizeof(timeout));
+#else
+            timeval tv;
+            tv.tv_sec = static_cast<long>(read_timeout_sec_);
```
Comment (yhirose) [inline httplib.h:5275]: Should use `write_timeout_sec`.

**Comment by yhirose** [comment]
@qq53, I discovered more problems. Did you really test your change? I am just curious...

**Comment by qq53** [comment]
it's my fault, I forgot to edit it.

**Comment by yhirose** [comment]
I am not blaming you at all. :) I was just asking if you really tested this pull request.

**Comment by qq53** [comment]
yeah,i test it on linux and windows.because request always return timeout on windows,so i debug the code,and find linux use struct to set value,but windows use dword in msdn document

**Pull request merged**